Mathlib/Combinatorics/SimpleGraph/LapMatrix.lean

```diff
@@ -0,0 +1,207 @@
+    refine Or.inl <| sum_eq_zero λ i _ ↦ (sum_eq_zero λ j _ ↦ ?_)
+    simpa only [ite_eq_right_iff, zero_lt_two, pow_eq_zero_iff, sub_eq_zero] using h i j
+
+theorem lapMatrix_toLinearMap₂'_zero_iff_adj (x : V → ℝ) :
+    Matrix.toLin' (G.lapMatrix ℝ) x = 0 ↔ ∀ i j : V, G.Adj i j → x i = x j := by
+  rw [← (isPosSemidef_lapMatrix ℝ G).toLinearMap₂'_zero_iff x, star_trivial,
+      lapMatrix_toLinearMap₂'_apply'_zero_iff_adj]
+
+theorem lapMatrix_toLinearMap₂'_apply'_zero_iff_reachable (x : V → ℝ) :
+    Matrix.toLinearMap₂' (G.lapMatrix ℝ) x x = 0 ↔ ∀ i j : V, G.Reachable i j → x i = x j := by
+  rw [lapMatrix_toLinearMap₂'_apply'_zero_iff_adj]
+  apply Iff.intro
+  · intro h i j
+    unfold Reachable
+    rw [Nonempty.forall]
+    intro w
+    induction' w with w i j _ hA _ h'
+    · rfl
+    · exact (h i j hA).trans h'
+  · exact fun h i j hA ↦ h i j (Adj.reachable hA)
+
+theorem ker_reachable_eq (x : V → ℝ) : Matrix.toLin' (G.lapMatrix ℝ) x = 0 ↔
+    ∀ i j : V, G.Reachable i j → x i = x j := by
+  rw [← (isPosSemidef_lapMatrix ℝ G).toLinearMap₂'_zero_iff, star_trivial,
+      lapMatrix_toLinearMap₂'_apply'_zero_iff_reachable]
+
+variable [DecidableEq G.ConnectedComponent]
+
+lemma lapMatrix_ker_basis_aux_well_defined {G : SimpleGraph V} [DecidableRel G.Adj]
+    [DecidableEq G.ConnectedComponent] (c : G.ConnectedComponent) :
+    (fun i ↦ if connectedComponentMk G i = c then 1 else 0) ∈
+      LinearMap.ker (toLin' (lapMatrix ℝ G)) := by
+  rw [LinearMap.mem_ker, ker_reachable_eq]
+  intro i j h
+  split_ifs with h₁ h₂ h₃
+  · rfl
+  · rw [← ConnectedComponent.eq] at h
+    exact (h₂ (h₁ ▸ h.symm)).elim
+  · rw [← ConnectedComponent.eq] at h
+    exact (h₁ (h₃ ▸ h)).elim
+  · rfl
+
+/-- Given a connected component `c` of a graph `G`, `lapMatrix_ker_basis_aux c` is the map
+  `V → ℝ` which is `1` on the vertices in `c` and `0` elsewhere.
+  The family of these maps indexed by the connected components of `G` proves to be a basis
+  of the kernel of `lapMatrix G R` -/
+def lapMatrix_ker_basis_aux (c : G.ConnectedComponent) :
+    LinearMap.ker (Matrix.toLin' (G.lapMatrix ℝ)) :=
+  ⟨fun i ↦ if G.connectedComponentMk i = c then (1 : ℝ)  else 0,
+    lapMatrix_ker_basis_aux_well_defined c⟩
+
+lemma lapMatrix_ker_basis_aux_linearIndependent :
+    LinearIndependent ℝ (lapMatrix_ker_basis_aux G) := by
+  rw [Fintype.linearIndependent_iff]
+  intro g h0
+  rw [Subtype.ext_iff] at h0
+  have h : ∑ c, g c • lapMatrix_ker_basis_aux G c = fun i ↦ g (connectedComponentMk G i)
+  · simp only [lapMatrix_ker_basis_aux, SetLike.mk_smul_mk, AddSubmonoid.coe_finset_sum]
+    conv_lhs => enter [2, c, j]; rw [Pi.smul_apply, smul_eq_mul, mul_ite, mul_one, mul_zero]
+    ext i
+    simp only [Finset.sum_apply, sum_ite_eq, mem_univ, ite_true]
+  rw [h] at h0
+  intro c
+  obtain ⟨i, h'⟩ : ∃ i : V, G.connectedComponentMk i = c := Quot.exists_rep c
+  exact h' ▸ congrFun h0 i
+
+lemma lapMatrix_ker_basis_aux_spanning :
```

Comment:
Mathlib names always follow the syntax of the lemma, not the meaning of it, unless "spanning" is a convention specified somewhere and I'm not familiar with it.
```suggestion
lemma top_le_span_range_lapMatrix_ker_basis_aux :
```